sensor: drivers: Panasonic sngcja5 particle matter sensor

pideu-sj opened the pull request:
Support for Panasonic's [SN-GCJA5](https://industry.panasonic.eu/products/components/sensors/particulate-matter-sensor) Particle Matter Sensor.
Added sensor drivers and sensor sample for example usage.

pideu-sj committed:
sensor: driver: Add support for Panasonic SN-GCJA5 laser particle sensor

Added driver for Panasonic's "particulate matters (PM) sensor" laser type.
PM sensor is able to detect smoke, environmental dust and other
unwanted dangerous pollutant in the air that surrounds us.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj committed:
sensor: sample: Add sample for Panasonic SN-GCJA5 laser particle sensor

Added sample for Panasonic's "particulate matters (PM) sensor" laser type.
The sample detects values for PM1, PM2.5, PM10 and the particle counts
periodically and prints them once a second to the console.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,179 @@
+	return 0;
+}
+
+static const struct sensor_driver_api sngcja5_driver_api = {
```
I'm always getting following error message in the twister runs: 
`zephyr/drivers/sensor/sngcja5/sngcja5.c:163:39: error: 'sngcja5_driver_api' defined but not used [-Werror=unused-const-variable=]`

Can someone explain me what I'm doing wrong, or what I have to do to pass?

MaureenHelm replied:
There probably aren't any enabled instances of `panasonic,sngcja5` in your devicetree

pideu-sj replied:
But I've added the _compatible = "panasonic,sngcja5";_ to my _arduino_i2c.overlay_ file. So i thought that it has to be in the devicetree afterwards?

MaureenHelm replied:
is `status = okay` ? do you see the node in the build artifact `zephyr.dts`? if you don't, that would suggest you're not including the overlay in the build

pideu-sj replied:
Yes I see the node in the build artifact zephyr.dts. Don't getting this error message in the twister runs anymore.

MaureenHelm commented on drivers/sensor/sngcja5/Kconfig:
```diff
@@ -0,0 +1,11 @@
+config SNGCJA5
+	bool "SN-GCJA5 laser particle sensor"
+	default y
+	depends on I2C
```
```suggestion
	depends on DT_HAS_PANASONIC_SNGCJA5_ENABLED`
	select I2C
```

pideu-sj replied:
Pushed as requested in latest pull request update.

MaureenHelm commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,179 @@
+
+LOG_MODULE_REGISTER(SNGCJA5, CONFIG_SENSOR_LOG_LEVEL);
+
+static int read_bytes(const struct i2c_dt_spec *dev, uint8_t addr, uint8_t *data,
+		      uint32_t num_bytes)
+{
+	struct i2c_msg msgs[2];
+
+	msgs[0].buf = &addr;
+	msgs[0].len = 1U;
+	msgs[0].flags = I2C_MSG_WRITE;
+
+	msgs[1].buf = data;
+	msgs[1].len = num_bytes;
+	msgs[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;
+
+	return i2c_transfer_dt(dev, &msgs[0], 2);
+}
```
Just use `i2c_burst_read_dt` instead

pideu-sj replied:
Pushed as requested in latest pull request update.

MaureenHelm commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,179 @@
+{
+	const struct sngcja5_config *config = dev->config;
+
+	if (!device_is_ready(config->i2c.bus)) {
```
`i2c_is_ready_dt`

pideu-sj replied:
Pushed as requested in latest pull request update.

MaureenHelm commented on dts/bindings/sensor/panasonic,sngcja5.yaml:
```diff
@@ -0,0 +1,8 @@
+
+compatible: "panasonic,sngcja5"
+
+include: i2c-device.yaml
```
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

MaureenHelm commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,179 @@
+	int32_t uval;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
+	if (chan != SENSOR_CHAN_ALL) {
```
Need to use a more specific channel type. If there isn't one in `enum sensor_channel` that makes sense for this sensor, then consider defining a driver specific channel. See drivers that use `SENSOR_CHAN_PRIV_START` for example

MaureenHelm commented on samples/sensor/sngcja5/src/main.c:
```diff
@@ -0,0 +1,74 @@
+void main(void)
+{
+	int ret;
+	const struct device *const dev = DEVICE_DT_GET_ONE(panasonic_sngcja5);
```
Please make this sample more generic so it can support other particle sensor drivers. See `samples/sensor/accel_polling` for example

pideu-sj replied:
Did that on last pull request update.

pideu-sj commented:
Hello @MaureenHelm 
sorry was caught on other projects on last weeks, but will go on here in the next days. Could you please remove the stale label?

pideu-sj commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,171 @@
+		return -EIO;
+	}
+
+	*value = (buf[1] << 8) | buf[0];
```
When using `sys_be16_to_cpu` like in following snippet:
```
static int read_register_2(const struct i2c_dt_spec *dev, uint8_t addr, uint16_t *value)
{
	if (i2c_burst_read_dt(dev, addr, (uint8_t *)value, sizeof(*value)) < 0) {
		LOG_ERR("i2c_burst_read_dt() @ i2c failed");
		return -EIO;
	}

	*value = sys_be16_to_cpu(*value);

	return 0;
}
```
Like it's done for example in `drivers/sensor/tmp112/tmp112.c`, i would get the wrong byte order for my read data. Should be: 
``
new sample:
pc0_5.1: 0x0003
``
But is:
``
new sample:
pc0_5.1: 0x0300
``
So i guess i have to stick to the previous solution.

kartben replied:
These are actually little-endian, not big-endian, so using `*_le*.` versions of the conversion macros should work as expected

pideu-sj commented on drivers/sensor/sngcja5/sngcja5.c:
```diff
@@ -0,0 +1,171 @@
+		return -EIO;
+	}
+
+	*value = (buf[3] << 24) | (buf[2] << 16) | (buf[1] << 8) | buf[0];
```
See https://github.com/zephyrproject-rtos/zephyr/pull/55765/commits/755d6357018471d9dfe3ae15d5a7d58dba58bd12#r1212827727

pideu-sj commented on include/zephyr/drivers/sensor.h:
```diff
@@ -101,6 +101,16 @@ enum sensor_channel {
+	 * 5. (5.0-7.5µm)
+	 * 6. (7.5-10.0µm)
+	 */
+	SENSOR_CHAN_PARTICLE_COUNT,
```
Updated comment

pideu-sj commented on include/zephyr/drivers/sensor.h:
```diff
@@ -351,47 +361,40 @@ typedef void (*sensor_trigger_handler_t)(const struct device *dev,
-				 enum sensor_attribute attr,
-				 const struct sensor_value *val);
+typedef int (*sensor_attr_set_t)(const struct device *dev, enum sensor_channel chan,
+				 enum sensor_attribute attr, const struct sensor_value *val);
```
Did this on last pull-request update.

MaureenHelm commented on drivers/sensor/Kconfig:
```diff
@@ -85,6 +85,101 @@ source "drivers/sensor/dht/Kconfig"
+source "drivers/sensor/nrf5/Kconfig"
+source "drivers/sensor/opt3001/Kconfig"
+source "drivers/sensor/pms7003/Kconfig"
+source "drivers/sensor/qdec_mcux/Kconfig"
+source "drivers/sensor/qdec_nrfx/Kconfig"
+source "drivers/sensor/qdec_sam/Kconfig"
+source "drivers/sensor/qdec_stm32/Kconfig"
+source "drivers/sensor/sbs_gauge/Kconfig"
+source "drivers/sensor/sgp40/Kconfig"
+source "drivers/sensor/shtcx/Kconfig"
+source "drivers/sensor/sht3xd/Kconfig"
+source "drivers/sensor/sht4x/Kconfig"
+source "drivers/sensor/si7006/Kconfig"
+source "drivers/sensor/si7055/Kconfig"
+source "drivers/sensor/si7060/Kconfig"
+source "drivers/sensor/si7210/Kconfig"
+source "drivers/sensor/sm351lt/Kconfig"
+source "drivers/sensor/sngcja5/Kconfig"
+source "drivers/sensor/stm32_temp/Kconfig"
+source "drivers/sensor/stm32_vbat/Kconfig"
+source "drivers/sensor/stts751/Kconfig"
+source "drivers/sensor/sx9500/Kconfig"
+source "drivers/sensor/th02/Kconfig"
+source "drivers/sensor/tmp007/Kconfig"
+source "drivers/sensor/tmp108/Kconfig"
+source "drivers/sensor/tmp112/Kconfig"
+source "drivers/sensor/tmp116/Kconfig"
+source "drivers/sensor/vcnl4040/Kconfig"
+source "drivers/sensor/vl53l0x/Kconfig"
+source "drivers/sensor/vl53l1x/Kconfig"
+source "drivers/sensor/nxp_kinetis_temp/Kconfig"
+source "drivers/sensor/wsen_hids/Kconfig"
+source "drivers/sensor/wsen_itds/Kconfig"
+source "drivers/sensor/wsen_tids/Kconfig"
+source "drivers/sensor/mcux_acmp/Kconfig"
+source "drivers/sensor/nuvoton_tach_npcx/Kconfig"
+source "drivers/sensor/nuvoton_adc_cmp_npcx/Kconfig"
+source "drivers/sensor/ite_tach_it8xxx2/Kconfig"
+source "drivers/sensor/ite_vcmp_it8xxx2/Kconfig"
+source "drivers/sensor/pcnt_esp32/Kconfig"
```
Looks like there were some rebasing problems

pideu-sj replied:
Oh Yeah. Not sure how that happened. Fixed in pull-request update. Thanks!

pideu-sj commented on drivers/sensor/CMakeLists.txt:
```diff
@@ -112,6 +112,8 @@ add_subdirectory_ifdef(CONFIG_SI7055 si7055)
 add_subdirectory_ifdef(CONFIG_SI7210 si7210)
 add_subdirectory_ifdef(CONFIG_SM351LT sm351lt)
+add_subdirectory_ifdef(CONFIG_SNGCJA5 sngcja5)
+add_subdirectory_ifdef(CONFIG_HAS_STMEMSC	stmemsc)
```
You are right. Again not sure, how that made it into the commit. Fixed that in newest update. Thanks.

MaureenHelm reviewed (commented):
Please rename the sample and update the documentation to reflect the change making it generic to any particle sensor driver

pideu-sj commented on include/zephyr/drivers/sensor.h:
```diff
@@ -106,6 +106,16 @@ enum sensor_channel {
 	/** Altitude, in meters */
 	SENSOR_CHAN_ALTITUDE,
 
+	/**
+	 * Particle count array
+	 * value1[(0.3-0.5µm)],
+	 * value2[(0.5-1.0µm)],
+	 * value3[(1.0-2.5µm)],
+	 * value4[(2.5-5.0µm)],
+	 * value5[(5.0-7.5µm)],
+	 * value6[(7.5-10.0µm)]
+	 */
+	SENSOR_CHAN_PARTICLE_COUNT,
```
I've found sensors like the `DFRobot SEN0460` that has the same 6 values available (https://wiki.dfrobot.com/Gravity_PM2.5_Air_Quality_Sensor_SKU_SEN0460). But have also found sensor like the `SPS30` that just provides 5 count values (https://sensirion.com/media/documents/8600FF88/616542B5/Sensirion_PM_Sensors_Datasheet_SPS30.pdf).

avisconti replied:
shouldn't this change be in a separate commit?

kartben replied:
At a minimum this should be a different commit, yes. 

But I am -1 on adding this new "array" channel as both your linked datasheets actually show that different sensors will likely use different "classes" of particle counts in this array i.e PM 0.5/1.0/2.5/4.0/10.0 for SPS30, and PM 0.3/0.5/1.0/2.5/5.0/10.0, none of which correspond to the ranges proposed here. HM3300/3600 sensors have yet another range (https://files.seeedstudio.com/wiki/Grove-Laser_PM2.5_Sensor-HM3301/res/HM-3300%263600_V2.1.pdf) and so on.

What would probably work though is adding `SENSOR_CHAN_PC_{X_Y}` channels.

pideu-sj replied:
@kartben 
Thanks a lot for taking over and reviewing.

On first step I spent an own commit for this change, like you've requested.

And you are right, it's difficult, to generalize the particle count values for several sensors.

> What would probably work though is adding SENSOR_CHAN_PC_{X_Y} channels.

Could you give an example, how this would look like for my sngcja5 sensor with the PM 0.5/1.0/2.5/5.0/7.5/10.0 ? Haven't seen it in that way here before?

kartben replied:
> > What would probably work though is adding SENSOR_CHAN_PC_{X_Y} channels.
> 
> Could you give an example, how this would look like for my sngcja5 sensor with the PM 0.5/1.0/2.5/5.0/7.5/10.0 ? Haven't seen it in that way here before?

I mean just like you currently have `SENSOR_CHAN_PM_1_0`, `SENSOR_CHAN_PM_2_5`, etc. there could also be new channels for `SENSOR_CHAN_PC_0_3`, `SENSOR_CHAN_PC_0_5`, etc.?
Note that something like #63830 would probably make things easier here, but in the meantime I think adding channels for the common particle count sizes would work, but I will let @MaureenHelm and other core contributors to the sensor API confirm.

teburd replied:
Please go with this route today

yperess replied:
I'm not really seeing the need for this to be a part of sensor.h, why can't this highly specific 6 value channel be added in sngcja5.h? I would also like to add that with the new async API you're able to have an index along with a channel which would service this model much better. SENSOR_CHAN_PARTICLE_COUNT would be used along with an index to provide any number of dimensions to the measurement. We would still need a way to describe what each index means.

pideu-sj commented:
> Please rename the sample and update the documentation to reflect the change making it generic to any particle sensor driver

Done in last update. Changed naming and documentation from sample `sngcja5` to `particulate_matter`.

pideu-sj commented:
@MaureenHelm 

Did you already found some time to review?

pideu-sj commented:
Hi @galak @avisconti @teburd , 

can maybe someone else review?

avisconti reviewed (dismissed):
LGTM (with a small non blocking comment, though)

pideu-sj commented:
@MaureenHelm 
could we maybe reopen this PR and remove the stale label?

It's been a while, but in comment https://github.com/zephyrproject-rtos/zephyr/pull/55765#issuecomment-1579962884 I said, that last requested changes from you site are added. So can we merge this PR now? @avisconti already gave the approval in the meantime.

pideu-sj commented:
Hi @teburd and @avisconti
had to rebase to match the compliance check and update manifest to pass all checks after removal of stale label.

pideu-sj commented:
Hi @MaureenHelm, @avisconti,
could you maybe find some minuets to check if we can merge the PR?
Thanks a lot.

pideu-sj commented:
Hi to all listed reviewers,
if @MaureenHelm and @avisconti are not available, could someone else maybe review?
Thanks a lot!

pideu-sj commented:
Hello to the reviewers in the list @avisconti, @nashif, @galak, @yperess, @kartben, @tristan-google,
the Pull-Request already got two times the stale label. Can maybe someone else review if @MaureenHelm is not available?
Thanks a lot!

pideu-sj committed:
include: zephyr: drivers: sensor: Add particle count array sensor channel

Particle matter sensors like the SNGCJA5 are able to detect particle
counts. This sensor channel array needs to be added.

Signed-off-by: Steffen Jahnke <steffen.jahnke@eu.panasonic.com>

pideu-sj commented:
@MaureenHelm and other core contributors to the sensor API,
could you confirm, that I should change sensor channels for particle count, like @kartben suggested here: https://github.com/zephyrproject-rtos/zephyr/pull/55765/files#r1391162757
Thanks a lot.

yperess commented on samples/sensor/particulate_matter/sample.yaml:
I'm going to let @MaureenHelm decide the path on this, but I'm not seeing a strong benefit to adding yet another sensor specific sample to sensors. The sensor shell sample can do this just fine.

pideu-sj replied:
@yperess, thanks for reviewing. 
Okay, so I will wait for decision of @MaureenHelm, before I change this in the driver.
